### PR TITLE
ath79: add support for KuWFi N650 CPE

### PR DESCRIPTION
**Pull request by andyboeh**
The KuWFi N650 is a 5GHz outdoor wireless bridge based on QCA9563.

Specs
=====

CPU: QCA9563
RAM: 128MiB
Flash: 16MiB
Wireless: QCA9888 (5GHz only)
Ethernet: 2x GBit (via QCA8337)

Installation
============

From OEM firmware
-----------------

The OEM firmware has telnet enabled by default. If not, it can be enabled from the firmware web interface. You need a TFTP server on your computer and the OpenWrt factory image should be available as "n650factory.bin". It is assumed that your computer has the IP 192.168.1.1 and the N650 192.168.1.20 (default IP address).

1. Connect via Telnet to the device and log in with the default credentials "admin:admin"
2. Exploit the limited interface by typing "ps & /bin/sh"
3. Press <ENTER> to start the shell
4. Enter the following commands:
```
$ cd /tmp
$ tftp -r n650factory.bin -g 192.168.1.1
$ cat << EOF > /tmp/openwrt.sh
#!/bin/sh

IMAGE_NAME="\$1"

if [ ! -e \${IMAGE_NAME} ]; then
  echo "Image file not found: \${IMAGE_NAME}"
  exit 1
fi

. /usr/sbin/common.sh

kill_remaining TERM
sleep 3
kill_remaining KILL

run_ramfs mtd write \${IMAGE_NAME} firmware
sleep 2
reboot -f
EOF

$ chmod +x /tmp/openwrt.sh
$ /tmp/openwrt.sh n650factory.bin
```
Once the device reboots, it should load OpenWrt.

From UART
---------

UART installation is possible since the serial header is already soldered on. The pinout is GND - Tx - Rx - VCC from top to bottom (RJ45 ports are at the bottom). Connect with 115200 8N1.

First, boot OpenWrt from TFTP. Enter the following commands in the U-Boot shell, assuming your computer has the IP address 192.168.1.1 and a TFTP server running where the initramfs image is provided as n650.bin:
```
  setenv ipaddr 192.168.1.20
  setenv serverip 192.168.1.1
  tftpboot 0x84000000 n650.bin
  bootm
```
Once booted, transfer -loader.bin and -sysupgrade.bin images to the device at /tmp. Enter the following commands, replacing the filenames:
```
  mtd write /tmp/loader.bin loader
  sysupgrade /tmp/sysupgrade.bin
```
Reboot and OpenWrt should load from flash.

### CODE REVIEW

File: target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac [hunk 1]
```diff
@@ -78,6 +78,9 @@ case "$board" in
 		[ "$PHYNBR" -eq 1 ] && \
 			macaddr_add "$(mtd_get_mac_ascii u-boot-env ethaddr)" 1 > /sys${DEVPATH}/macaddress
 		;;
+	kuwfi,n650)
```
Comment (neheb) [inline target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:81]: You can do this in dts with Mac-base

Comment (andyboeh) [reply target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:81]: Hm, didn't work for me. Could you share an example?

Comment (neheb) [reply target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:81]: https://github.com/openwrt/openwrt/commit/b7f26c63929da908f83c767689bc25ef161db466

Comment (andyboeh) [reply target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:81]: Thanks, I'll fix that. I was probably too tired when I crafted the DTS.

File: target/linux/ath79/dts/qca9563_kuwfi_n650.dts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+			0x04 0x00080080 /* PORT0 PAD MODE CTRL */
+			0x7c 0x0000007e /* PORT0_STATUS */
+		>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	nvmem-cells = <&macaddr_art_0>;
```
Comment (neheb) [inline target/linux/ath79/dts/qca9563_kuwfi_n650.dts:163]: this needs a 0 at the end. Otherwise it won't work.

Comment (andyboeh) [reply target/linux/ath79/dts/qca9563_kuwfi_n650.dts:163]: You're of course right. I didn't see the changed mac address of the ethernet interface. I'm quite certain, though, that it worked before adding the offset to WiFi.

Comment (andyboeh) [reply target/linux/ath79/dts/qca9563_kuwfi_n650.dts:163]: Ah, got it: it worked because I didn't have the "mac-base" compatible string!

**Comment by neheb** [comment]
@olemmela should this be converted to DSA?

**Comment by andyboeh** [comment]
Is ath79 finally being converted to DSA? I'd love to do so, since all my other devices are DSA.

**Comment by hauke** [comment]
We should not switch random ath79 devices to DSA

**Comment by andyboeh** [comment]
@ynezz Seeing that you backported the TP-Link CPE710v2 to 24.10, are you going to do the same for this device or shall I create a backport PR?

**Comment by Neustradamus** [comment]
@andyboeh: Good job!

**Pull request merged**